lib/Sema/CSSimplify.cpp

```diff
@@ -2928,10 +2928,17 @@ ConstraintSystem::matchTypes(Type type1, Type type2, ConstraintKind kind,
   auto desugar1 = type1->getDesugaredType();
   auto desugar2 = type2->getDesugaredType();
 
-  // If the types are obviously equivalent, we're done.
-  if (desugar1->isEqual(desugar2) &&
-      !isa<InOutType>(desugar2))
-    return getTypeMatchSuccess();
+  // If both sides are dependent members without type variables, it's
+  // possible that base type is incorrect e.g. `U.Element` where `U`
+  // is a generic parameter, so checking equality here would lead to
+  // incorrect behavior, let's defer it until later proper check.
+  if (!(desugar1->is<DependentMemberType>() &&
```

Comment:
Is the base type a generic parameter or concrete in this case? I'm a bit surprised to see DependentMemberTypes showing up here at all, since usually they only appear in interface types and constraint systems use contextual types?

Reply:
That's a consequence of `simplifyTypeImpl` implementation, it would re-create `DependentMemberType` if the base was incorrect concrete type. I think at some point we'd want to change that behavior but it could be done separately. It mostly shows up in diagnostics when missing conformances are allowed so for signatures like that `<T: Sequence, U: Sequence, T.Element == U.Element>` if we try `T = Int` and `U = String` for example, we'll get a requirement constraint as `Int.Element == String.Element` which used to be considered unsolved but now is going to fail.

Reply:
Oh, right. You might want to update the comment then -- the base will never be a generic parameter, but it might be a concrete type that does not conform to U.

I guess we have to allow this because simplifyType() cannot report failure, but I imagine if this happens then a conformance constraint somewhere else should have failed, correct?

Reply:
Yes, that's correct, it's never going to show up during normal solving, only in diagnostic mode and if conformance constraint fails.

Reply:
Sorry, comment should have been more clear about relationship between this problem and generic parameters.

Reply:
Updated comment to talk about concrete types instead.